File: roles/postfix/files/monitoring/check_smtp/php/send.php
#!/usr/bin/php
<?php

$self = array_shift($argv);
$domain = array_shift($argv);
if(!preg_match('/^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$/i', $domain)) die("bad domain: $domain\n");

$dir = dirname(__FILE__);

$conf = [];
$lines = file_get_contents("$dir/check.conf");
if($lines === false) die("could not open config: $dir/check.conf\n");
foreach(array_map('trim', explode("\n", $lines)) as $line) {
    if(!preg_match('/^([a-z][a-z0-9_\.]*)\s+(.+)$/i', $line, $m)) continue;
    $conf[$m[1]] = $m[2];
}

foreach(['smtp_out', 'address', 'auth_secret', 'source'] as $c) {
    if(!array_key_exists($c, $conf)) die("missing config parameter: $c\n");
}

$from = 'listmaster@'.$domain;
$to = $conf['address'].'@'.$domain;
$now = time();
$source = $conf['source'];

$signature = hash_hmac('sha1', $now.'@'.$source, $conf['auth_secret']);

$headers = [
    "X-Source: $source",
    "X-Generated: $now",
    "X-Authentication: $signature",
];

if(!mail($to, 'SMTP chain check', 'This is a SMTP chain check message', implode("\n", $headers)))
    die("could not send message\n");

File: roles/postfix/files/monitoring/check_smtp/sympa/send.pl
#!/usr/bin/perl

use File::Basename;
use File::Slurp;
use Net::SMTP;
use Digest::HMAC_SHA1 qw(hmac_sha1_hex);

my $domain = shift @ARGV;
die "bad domain: $domain\n" unless $domain =~ /^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$/i;

my $dir = dirname(__FILE__);

my %conf;
foreach my $line (read_file("$dir/check.conf", chomp => 1)) {
    next unless $line =~ /^([a-z][a-z0-9_\.]*)\s+(.+)$/i;
    $conf{$1} = $2;
}

foreach my $c ('smtp_out', 'address', 'auth_secret', 'source') {
    die "missing config parameter: $c\n" unless defined $conf{$c};
}

my $from = 'listmaster@'.$domain;
my $to = $conf{'address'}.'@'.$domain;
my $now = time;
my $source = $conf{'source'};

my $signature = hmac_sha1_hex($now.'@'.$source, $conf{'auth_secret'});

my $smtp = Net::SMTP->new($conf{'smtp_out'});
$smtp->mail($from);

if($smtp->to($to)) {
    $smtp->data();
    $smtp->datasend("To: $to\n");
    $smtp->datasend("X-Source: $source\n");
    $smtp->datasend("X-Generated: $now\n");
    $smtp->datasend("X-Authentication: $signature\n");
    $smtp->datasend("Subject: SMTP chain check\n");
    $smtp->datasend("\n");
    $smtp->datasend("This is a SMTP chain check message\n");
    $smtp->dataend();
} else {
    die "smtp error: ".$smtp->message()."\n";
}

$smtp->quit;

0;
